ImperatorToCK3/Source/Imperator/Characters/Character.cpp

```diff
@@ -71,32 +49,8 @@ void ImperatorWorld::Character::registerKeys()
 		family = std::pair(familyInt.getInt(), nullptr);
 	});
 	registerKeyword("dna", [this](const std::string& unused, std::istream& theStream) {
-		const commonItems::singleString dnaStr(theStream);
-		dna = dnaStr.getString();
-		if (dna.size() == 552)
-		{
-			const std::string& decodedDnaStr = base64_decode(dna);
-
-			std::string binary_outputInformations;
-			for (std::size_t i = 0; i < decodedDnaStr.size(); ++i)
-			{
-				std::bitset<8> b(decodedDnaStr.c_str()[i]);
-				binary_outputInformations += b.to_string();
-			}
-
-			//hair
-			portraitData.hairColorPaletteCoordinates.x = unsigned int (binaryToDecimal(stoll(binary_outputInformations.substr(0, 18))) / 512);
-			portraitData.hairColorPaletteCoordinates.y = unsigned int (binaryToDecimal(stoll(binary_outputInformations.substr(0, 18))) % 512);
-			//skin
-			portraitData.skinColorPaletteCoordinates.x = unsigned int(binaryToDecimal(stoll(binary_outputInformations.substr(5, 18))) / 512);
-			portraitData.skinColorPaletteCoordinates.y = unsigned int(binaryToDecimal(stoll(binary_outputInformations.substr(5, 18))) % 512);
-			//eyes
-			portraitData.eyeColorPaletteCoordinates.x = unsigned int(binaryToDecimal(stoll(binary_outputInformations.substr(10, 18))) / 512);
-			portraitData.eyeColorPaletteCoordinates.y = unsigned int(binaryToDecimal(stoll(binary_outputInformations.substr(10, 18))) % 512);
-			Log(LogLevel::Debug) << "Char ID " << charID << " hair color palette X Y coordinates: " << portraitData.hairColorPaletteCoordinates.x << " " << portraitData.hairColorPaletteCoordinates.y; // debug
-			Log(LogLevel::Debug) << "Char ID " << charID << " skin color palette X Y coordinates: " << portraitData.skinColorPaletteCoordinates.x << " " << portraitData.skinColorPaletteCoordinates.y; // debug
-			Log(LogLevel::Debug) << "Char ID " << charID << " eye color palette X Y coordinates: " << portraitData.eyeColorPaletteCoordinates.x << " " << portraitData.eyeColorPaletteCoordinates.y; // debug
-		}
+		dna = commonItems::singleString(theStream).getString();
+		if (dna && dna.value().size() == 552) portraitData.emplace(CharacterPortraitData(dna.value()));
```

Comment:
dna here has been assigned a string class and will always be true. You're probably asking if !*dna.empty()

Reply:
In that case I think `if (dna && dna.value().size() == 552)` is enough.